### PR TITLE
Simplify construction of parameteric pw affs

### PR DESCRIPTION
**Pull request by tobiasgrosser**
This pull request proposes three new functions to simplify the generation of pw_aff expressions on parameter spaces. All expressions are generated on a universe domain.

```
#include <isl/aff.h>                                                     
__isl_give isl_pw_aff *isl_pw_aff_param_from_id(__isl_take isl_id *id);  
__isl_give isl_pw_aff *isl_pw_aff_val_from_val(__isl_take isl_val *val); 
__isl_give isl_pw_aff *isl_pw_aff_val_from_si(__isl_keep isl_ctx *ctx, int val);        
```

### CODE REVIEW

File: doc/user.pod [hunk 1]
```diff
@@ -3072,8 +3072,8 @@ created using the following functions.
 		__isl_take isl_set *set,
 		__isl_take isl_qpolynomial *qp);
 
-The following convenience functions first create a base expression and
-then create a piecewise expression over a universe domain.
+The following convenience functions first create a base expression and then
+create a piecewise expression over a universe domain of a given (local) space.
```
Comment (ftynse) [inline doc/user.pod]: I'm not sure that "over a universe domain of a space" is mathematically correct.  "over domain that is a universe set in the given (local) space", leaving that to @skimo-openhub.

File: doc/user.pod [hunk 2]
```diff
@@ -3099,6 +3099,14 @@ then create a piecewise expression over a universe domain.
 	__isl_give isl_pw_qpolynomial *isl_pw_qpolynomial_zero(
 		__isl_take isl_space *space);
 
+The following convenience functions first create a base expression and then
+create a piecewise expression over a universe domain of a parameter space
+which contains exactly one parameter dimension corresponding to the newly
```
Comment (ftynse) [inline doc/user.pod]: "newly introduced" -> "given" ?

File: doc/user.pod [hunk 3]
```diff
@@ -3099,6 +3099,14 @@ then create a piecewise expression over a universe domain.
...
+introduced parameter dimensions.
```
Comment (ftynse) [inline doc/user.pod]: "dimension" (singular)

File: isl_aff.c [hunk 4]
```diff
@@ -216,6 +216,26 @@ __isl_give isl_aff *isl_aff_var_on_domain(__isl_take isl_local_space *ls,
 	return NULL;
 }
 
+/* Return a piecewise affine expression defined over a one dimensional
```
Comment (ftynse) [inline isl_aff.c]: "one-dimensional"

File: isl_test.c [hunk 5]
```diff
@@ -4763,6 +4763,33 @@ static isl_bool union_pw_aff_plain_is_equal(__isl_keep isl_union_pw_aff *upa,
...
+
+	equal = isl_pw_aff_plain_is_equal(pwa, pwa2);
+
+	isl_pw_aff_free(pwa);
+	isl_pw_aff_free(pwa2);
+
+	if (equal < 0)
+		return isl_stat_error;
+	if (!equal)
+		isl_die(ctx, isl_error_unknown, "can not construct pw from id",
```
Comment (ftynse) [inline isl_test.c]: `pw_aff`

File: isl_test.c [hunk 6]
```diff
@@ -6048,6 +6075,8 @@ int test_aff(isl_ctx *ctx)
 	isl_aff *aff;
 	int zero, equal;
 
+	if (test_pwa(ctx) < 0)
```
Comment (ftynse) [inline isl_test.c]: I'd call this function `test_pa` for consistency with the `test_upa` below

File: isl_aff.c [hunk 7]
```diff
@@ -236,6 +236,21 @@ __isl_give isl_pw_aff *isl_pw_aff_param_from_id(__isl_take isl_id *id)
 	return isl_pw_aff_from_aff(aff);
 }
 
+/* Return a piecewise affine expression defined over a zero dimensional
```
Comment (ftynse) [inline isl_aff.c]: "zero-dimensional"

File: isl_test.c [hunk 8]
```diff
@@ -4765,10 +4765,12 @@ static isl_bool union_pw_aff_plain_is_equal(__isl_keep isl_union_pw_aff *upa,
 
 /* Basic tests on isl_pw_aff.
  *
- * In particular test the construction of a single-parameter pw_aff from an id.
+ * In particular test the construction of a single-parameter pw_aff from an id
```
Comment (ftynse) [inline isl_test.c]: or a _zero-parameter_ pw_aff from an isl val

File: isl_test.c [hunk 9]
```diff
@@ -4804,6 +4804,19 @@ static int test_pwa(isl_ctx *ctx) {
...
+
+	equal = isl_pw_aff_plain_is_equal(pwa, pwa2);
+
+	isl_pw_aff_free(pwa);
+	isl_pw_aff_free(pwa2);
+
+	if (equal < 0)
+		return isl_stat_error;
+	if (!equal)
+		isl_die(ctx, isl_error_unknown, "can not construct pw from val",
```
Comment (ftynse) [inline isl_test.c]: from signed integer

**Review by ftynse — COMMENTED** [review]
All commit messages have the same typo "piecewiese".  Individual comments are mostly typos and minor suggestions.

File: doc/user.pod (hunk 1 above)
Comment (tobiasgrosser) [reply doc/user.pod]: Done.

File: doc/user.pod (hunk 2 above)
Comment (tobiasgrosser) [reply doc/user.pod]: Done.

File: doc/user.pod (hunk 3 above)
Comment (tobiasgrosser) [reply doc/user.pod]: I used 'id' instead.

File: isl_test.c (hunk 6 above)
Comment (tobiasgrosser) [reply isl_test.c]: Done.

File: isl_test.c (hunk 5 above)
Comment (tobiasgrosser) [reply isl_test.c]: Used 'piecewise aff' to be more clear.

File: isl_test.c (hunk 9 above)
Comment (tobiasgrosser) [reply isl_test.c]: Done.

File: isl_aff.c (hunk 7 above)
Comment (tobiasgrosser) [reply isl_aff.c]: Done.

File: isl_test.c (hunk 8 above)
Comment (tobiasgrosser) [reply isl_test.c]: Done.

**Comment by tobiasgrosser** [comment]
@ftynse, thanks for the comments. I addressed them all in the latest version of this patch set.

I missed a high-level comment from you. Do you think the functions themselves and specifically the function names picked make sense?

**Comment by ftynse** [comment]
These functions make sense in context of further set/map construction.  I don't see other potential uses now, but may be overlooking something.

I did think about the names.  Existing aff construction functions use the pattern "<something>_on_domain", but here we don't have the domain.  Alternative names can be along the lines of "`isl_pw_aff_cst_on_empty_domain_val`", "`isl_pw_aff_cst_on_empty_domain_si`" and "`isl_pw_aff_param_on_singular_domain`", but they look too wordy.  Using the pattern "`<something>_from_<something>`" is sort of associated with conversion, like `map_from_multi_union_pw_aff`, and often exported as overloads or even constructors.  For me, a conversion from a constant to a constant function defined over zero-dimensional space sounds okay, not sure about @skimo-openhub.

**Comment by tobiasgrosser** [comment]
OK, so we basically have the following two naming schemes:

### Conversion-style naming

These functions transform a constant value / id into a constant function over a zero-dimensional space and are consequently conversion style constructors.
```
isl_pw_aff_param_from_id
isl_pw_aff_val_from_val
isl_pw_aff_val_from_si
```

### Explicit function names (Alex)
```
isl_pw_aff_param_on_singular_domain
isl_pw_aff_cst_on_empty_domain_val
isl_pw_aff_cst_on_empty_domain_si
```

I think we should use param/val, as the term 'constant' might refer both to a parameter or an integer value -- basically everything that does not depend on a variable.

I also don't think that `empty_domain` is the right name. I would guess, if we call get_domain() on these pw_aff and ask if the domain `is_empty()` the reply would always be false. Not sure what would be a better name. The term 'singleton' would refer to a set with exactly one element, but does not say anything about the dimensionality of the set. We could use sth like `on_param_domain` to capture both the fact that the result is a singleton and/or that the space is a parameter space. Now, this won't give information about the number of parameter dimensions. However, as they are anyhow auto-aligned, that might be fine. Maybe the following:

### Explicit function names (Tobias)
```
isl_pw_aff_param_on_param_domain
isl_pw_aff_val_on_param_domain_val
isl_pw_aff_val_on_param_domain_si
```
